File: apps/frontend/evals/chord-detection/datasets.ts
import { readFile, writeFile } from "node:fs/promises";
import { createRequire } from "node:module";
import path from "node:path";
import {
  downloadFile,
  encodeUrlPath,
  ensureDir,
  exists,
  fileStatFingerprint,
  listFiles,
  removeDir,
} from "./fs-utils";
import { normalizeChordLabel } from "./label-map";
import type {
  DatasetId,
  DatasetLoadResult,
  DatasetSkip,
  EvalCliOptions,
  EvalSample,
} from "./types";

const require = createRequire(import.meta.url);
const extractZip = require("extract-zip") as typeof import("extract-zip");

const HF_DATASET = "severyn-k/isolated-guitar-chords";
const HF_SPLIT_ROOT = "data/Test";
const GUITARSET_RECORD_API = "https://zenodo.org/api/records/3371780";
const GUITARSET_FILE_KEYS = ["annotation.zip", "audio_mono-mic.zip"] as const;

interface HuggingFaceTreeItem {
  path: string;
  type: "file" | "directory";
  size?: number;
}

interface ZenodoRecord {
  files: Array<{ key: string; size: number; links: { self: string } }>;
}

interface GuitarSetJams {
  annotations: Array<{
    namespace: string;
    data: Array<{ time: number; duration: number; value: string }>;
  }>;
}

export async function prepareDatasets(options: EvalCliOptions): Promise<void> {
  for (const datasetId of options.datasets) {
    if (datasetId === "isolated-guitar-chords") await prepareIsolatedDataset(options);
    if (datasetId === "guitarset") await prepareGuitarSet(options);
  }
}

export async function loadDataset(
  datasetId: DatasetId,
  options: EvalCliOptions,
): Promise<DatasetLoadResult> {
  if (datasetId === "isolated-guitar-chords") return loadIsolatedDataset(options);
  return loadGuitarSet(options);
}

async function prepareIsolatedDataset(options: EvalCliOptions): Promise<void> {
  const root = isolatedRoot(options.cacheRoot);
  if (options.refreshDatasets) await removeDir(root);
  await ensureDir(root);
  const apiUrl = `https://huggingface.co/api/datasets/${HF_DATASET}/tree/main/${HF_SPLIT_ROOT}?recursive=1`;
  const items = (await fetchJson(apiUrl)) as HuggingFaceTreeItem[];
  const files = items.filter((item) => item.type === "file" && item.path.endsWith(".wav"));
  let downloaded = 0;
  for (const item of files) {
    const label = item.path.split("/").at(-2) ?? "";
    if (!normalizeChordLabel(label)) continue;
    const target = path.join(root, item.path);
    const url = `https://huggingface.co/datasets/${HF_DATASET}/resolve/main/${encodeUrlPath(item.path)}`;
    await downloadFile(url, target, options.refreshDatasets);
    downloaded++;
  }
  await writeJson(path.join(root, "manifest.json"), {
    datasetId: "isolated-guitar-chords",
    source: `https://huggingface.co/datasets/${HF_DATASET}`,
    split: "Test",
    downloaded,
  });
}

async function prepareGuitarSet(options: EvalCliOptions): Promise<void> {
  const root = guitarSetRoot(options.cacheRoot);
  if (options.refreshDatasets) await removeDir(root);
  await ensureDir(root);
  const record = (await fetchJson(GUITARSET_RECORD_API)) as ZenodoRecord;
  const archivesDir = path.join(root, "archives");
  const extractedDir = path.join(root, "extracted");
  await ensureDir(archivesDir);
  for (const key of GUITARSET_FILE_KEYS) {
    const file = record.files.find((item) => item.key === key);
    if (!file) throw new Error(`GuitarSet Zenodo file missing: ${key}`);
    const archivePath = path.join(archivesDir, key);
    await downloadFile(file.links.self, archivePath, options.refreshDatasets);
    const extractTarget = path.join(extractedDir, key.replace(/\.zip$/, ""));
    const marker = path.join(extractTarget, ".extracted");
    if (options.refreshDatasets || !(await exists(marker))) {
      await removeDir(extractTarget);
      await ensureDir(extractTarget);
      await extractZip(archivePath, { dir: extractTarget });
      await writeFile(marker, new Date().toISOString());
    }
  }
  await writeJson(path.join(root, "manifest.json"), {
    datasetId: "guitarset",
    source: "https://zenodo.org/records/3371780",
    files: GUITARSET_FILE_KEYS,
  });
}

async function loadIsolatedDataset(options: EvalCliOptions): Promise<DatasetLoadResult> {
  const root = isolatedRoot(options.cacheRoot);
  const files = await listFiles(path.join(root, HF_SPLIT_ROOT), (filePath) =>
    filePath.toLowerCase().endsWith(".wav"),
  );
  const skipped = new Map<string, number>();
  const samples: EvalSample[] = [];
  for (const filePath of files) {
    const label = path.basename(path.dirname(filePath));
    const expectedChordId = normalizeChordLabel(label);
    if (!expectedChordId) {
      increment(skipped, "unsupported label");
      continue;
    }
    const relative = path.relative(root, filePath);
    samples.push({
      id: `isolated:${relative}`,
      datasetId: "isolated-guitar-chords",
      expectedChordId,
      label,
      audioPath: filePath,
      sourcePath: relative,
      startSec: 0,
      endSec: Number.POSITIVE_INFINITY,
      sampleFingerprint: await fileStatFingerprint(filePath),
      metadata: { split: "Test" },
    });
  }
  return { datasetId: "isolated-guitar-chords", samples, skipped: toSkips(skipped) };
}

async function loadGuitarSet(options: EvalCliOptions): Promise<DatasetLoadResult> {
  const root = guitarSetRoot(options.cacheRoot);
  const annotationRoot = path.join(root, "extracted", "annotation.zip".replace(/\.zip$/, ""));
  const audioRoot = path.join(root, "extracted", "audio_mono-mic.zip".replace(/\.zip$/, ""));
  const jamsFiles = await listFiles(annotationRoot, (filePath) => filePath.endsWith(".jams"));
  const audioFiles = await listFiles(audioRoot, (filePath) =>
    filePath.toLowerCase().endsWith(".wav"),
  );
  const audioByBase = indexAudioByJamsBase(audioFiles);
  const skipped = new Map<string, number>();
  const samples: EvalSample[] = [];

  for (const jamsPath of jamsFiles) {
    const base = path.basename(jamsPath, ".jams");
    if (options.guitarSetMode === "comp" && !base.endsWith("_comp")) {
      increment(skipped, "non-comp performance");
      continue;
    }
    const audioPath = audioByBase.get(base);
    if (!audioPath) {
      increment(skipped, "missing mono-mic audio");
      continue;
    }
    const jams = JSON.parse(await readFile(jamsPath, "utf8")) as GuitarSetJams;
    const simpleChordAnnotation = jams.annotations.filter(
      (annotation) => annotation.namespace === "chord",
    )[0];
    const performedChordAnnotation = jams.annotations.filter(
      (annotation) => annotation.namespace === "chord",
    )[1];
    if (!simpleChordAnnotation) {
      increment(skipped, "missing chord annotation");
      continue;
    }
    const audioFingerprint = await fileStatFingerprint(audioPath);
    for (let i = 0; i < simpleChordAnnotation.data.length; i++) {
      const simple = simpleChordAnnotation.data[i];
      if (!simple || simple.duration < 0.25) {
        increment(skipped, "too-short chord segment");
        continue;
      }
      const performed = performedChordAnnotation?.data[i];
      const expectedChordId =
        (performed ? normalizeChordLabel(performed.value) : null) ??
        normalizeChordLabel(simple.value);
      if (!expectedChordId) {
        increment(skipped, "unsupported label");
        continue;
      }
      const sourcePath = path.relative(root, jamsPath);
      const startSec = simple.time;
      const endSec = simple.time + simple.duration;
      samples.push({
        id: `guitarset:${base}:${i}:${startSec.toFixed(3)}:${endSec.toFixed(3)}:${expectedChordId}`,
        datasetId: "guitarset",
        expectedChordId,
        label: performed?.value ?? simple.value,
        audioPath,
        sourcePath,
        startSec,
        endSec,
        sampleFingerprint: `${audioFingerprint}:${sourcePath}:${i}:${simple.value}:${performed?.value ?? ""}`,
        metadata: {
          performance: base,
          chordIndex: i,
          simpleLabel: simple.value,
          performedLabel: performed?.value ?? null,
          mode: options.guitarSetMode,
        },
      });
    }
  }
  return { datasetId: "guitarset", samples, skipped: toSkips(skipped) };
}

function isolatedRoot(cacheRoot: string): string {
  return path.join(cacheRoot, "datasets", "isolated-guitar-chords");
}

function guitarSetRoot(cacheRoot: string): string {
  return path.join(cacheRoot, "datasets", "guitarset");
}

function indexAudioByJamsBase(audioFiles: string[]): Map<string, string> {
  const out = new Map<string, string>();
  for (const audioPath of audioFiles) {
    const base = path.basename(audioPath, path.extname(audioPath));
    out.set(base, audioPath);
    out.set(base.replace(/_mic$/, ""), audioPath);
    out.set(base.replace(/_mono-mic$/, ""), audioPath);
  }
  return out;
}

async function fetchJson(url: string): Promise<unknown> {
  const response = await fetch(url);
  if (!response.ok)
    throw new Error(`request failed ${response.status} ${response.statusText}: ${url}`);
  return response.json();
}

async function writeJson(filePath: string, value: unknown): Promise<void> {
  await ensureDir(path.dirname(filePath));
  await writeFile(filePath, `${JSON.stringify(value, null, 2)}\n`);
}

function increment(map: Map<string, number>, reason: string): void {
  map.set(reason, (map.get(reason) ?? 0) + 1);
}

function toSkips(map: Map<string, number>): DatasetSkip[] {
  return [...map.entries()]
    .map(([reason, count]) => ({ reason, count }))
    .sort((a, b) => a.reason.localeCompare(b.reason));
}

File: apps/backend/app/evals/chord_detection/datasets.py
from __future__ import annotations

import json
import math
from pathlib import Path
from typing import Any

from .label_map import normalize_chord_label

DATASET_IDS = ("isolated-guitar-chords", "guitarset")
HF_SPLIT_ROOT = Path("data") / "Test"


def load_dataset(dataset_id: str, *, cache_root: Path, guitarset_mode: str = "comp") -> dict:
    if dataset_id == "isolated-guitar-chords":
        return load_isolated_dataset(cache_root)
    if dataset_id == "guitarset":
        return load_guitarset(cache_root, guitarset_mode=guitarset_mode)
    raise ValueError(f"unknown dataset: {dataset_id}")


def load_isolated_dataset(cache_root: Path) -> dict:
    root = cache_root / "datasets" / "isolated-guitar-chords"
    split_root = root / HF_SPLIT_ROOT
    skipped: dict[str, int] = {}
    samples: list[dict[str, Any]] = []
    for file_path in sorted(split_root.rglob("*.wav")) if split_root.exists() else []:
        label = file_path.parent.name
        expected_chord_id = normalize_chord_label(label)
        if expected_chord_id is None:
            increment(skipped, "unsupported label")
            continue
        relative = file_path.relative_to(root).as_posix()
        samples.append(
            {
                "id": f"isolated:{relative}",
                "datasetId": "isolated-guitar-chords",
                "expectedChordId": expected_chord_id,
                "label": label,
                "audioPath": str(file_path),
                "sourcePath": relative,
                "startSec": 0,
                "endSec": math.inf,
                "sampleFingerprint": file_stat_fingerprint(file_path),
                "metadata": {"split": "Test"},
            }
        )
    return {"datasetId": "isolated-guitar-chords", "samples": samples, "skipped": to_skips(skipped)}


def load_guitarset(cache_root: Path, *, guitarset_mode: str) -> dict:
    root = cache_root / "datasets" / "guitarset"
    annotation_root = root / "extracted" / "annotation"
    audio_root = root / "extracted" / "audio_mono-mic"
    skipped: dict[str, int] = {}
    samples: list[dict[str, Any]] = []
    jams_files = sorted(annotation_root.rglob("*.jams")) if annotation_root.exists() else []
    audio_by_base = index_audio_by_jams_base(
        sorted(audio_root.rglob("*.wav")) if audio_root.exists() else []
    )
    for jams_path in jams_files:
        base = jams_path.stem
        if guitarset_mode == "comp" and not base.endswith("_comp"):
            increment(skipped, "non-comp performance")
            continue
        audio_path = audio_by_base.get(base)
        if audio_path is None:
            increment(skipped, "missing mono-mic audio")
            continue
        jams = json.loads(jams_path.read_text())
        chord_annotations = [
            annotation for annotation in jams.get("annotations", []) if annotation.get("namespace") == "chord"
        ]
        simple_annotation = chord_annotations[0] if len(chord_annotations) > 0 else None
        performed_annotation = chord_annotations[1] if len(chord_annotations) > 1 else None
        if simple_annotation is None:
            increment(skipped, "missing chord annotation")
            continue
        audio_fingerprint = file_stat_fingerprint(audio_path)
        simple_data = simple_annotation.get("data", [])
        performed_data = performed_annotation.get("data", []) if performed_annotation else []
        for index, simple in enumerate(simple_data):
            duration = float(simple.get("duration", 0))
            if duration < 0.25:
                increment(skipped, "too-short chord segment")
                continue
            performed = performed_data[index] if index < len(performed_data) else None
            simple_value = str(simple.get("value", ""))
            performed_value = str(performed.get("value", "")) if performed else ""
            expected_chord_id = (
                normalize_chord_label(performed_value) if performed else None
            ) or normalize_chord_label(simple_value)
            if expected_chord_id is None:
                increment(skipped, "unsupported label")
                continue
            source_path = jams_path.relative_to(root).as_posix()
            start_sec = float(simple.get("time", 0))
            end_sec = start_sec + duration
            samples.append(
                {
                    "id": f"guitarset:{base}:{index}:{start_sec:.3f}:{end_sec:.3f}:{expected_chord_id}",
                    "datasetId": "guitarset",
                    "expectedChordId": expected_chord_id,
                    "label": performed_value if performed else simple_value,
                    "audioPath": str(audio_path),
                    "sourcePath": source_path,
                    "startSec": start_sec,
                    "endSec": end_sec,
                    "sampleFingerprint": f"{audio_fingerprint}:{source_path}:{index}:{simple_value}:{performed_value}",
                    "metadata": {
                        "performance": base,
                        "chordIndex": index,
                        "simpleLabel": simple_value,
                        "performedLabel": performed_value if performed else None,
                        "mode": guitarset_mode,
                    },
                }
            )
    return {"datasetId": "guitarset", "samples": samples, "skipped": to_skips(skipped)}


def index_audio_by_jams_base(audio_files: list[Path]) -> dict[str, Path]:
    out: dict[str, Path] = {}
    for audio_path in audio_files:
        base = audio_path.stem
        out[base] = audio_path
        out[base.removesuffix("_mic")] = audio_path
        out[base.removesuffix("_mono-mic")] = audio_path
    return out


def file_stat_fingerprint(file_path: Path) -> str:
    stat = file_path.stat()
    return f"{stat.st_size}:{round(stat.st_mtime * 1000)}"


def increment(skipped: dict[str, int], reason: str) -> None:
    skipped[reason] = skipped.get(reason, 0) + 1


def to_skips(skipped: dict[str, int]) -> list[dict[str, int | str]]:
    return [{"reason": reason, "count": count} for reason, count in sorted(skipped.items())]
